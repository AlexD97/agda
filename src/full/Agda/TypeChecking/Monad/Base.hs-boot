{-# LANGUAGE TypeFamilies               #-} -- for type equality ~

module Agda.TypeChecking.Monad.Base where

import Control.Monad.IO.Class (MonadIO)
import Control.Monad.Trans          ( MonadTrans )
import Control.Monad.Trans.Control  ( MonadTransControl )
import Data.IORef (IORef)
import Data.Map (Map)

import Agda.Syntax.Common (Nat)
import Agda.Syntax.Internal (Dom, Name, Term, Type)
import Agda.Syntax.Concrete.Name (TopLevelModuleName)
import Agda.Utils.FileName (AbsolutePath)

data Warning

data TCErr
data TCWarning
data NamedMeta
data HighlightingMethod
instance Show HighlightingMethod
instance Read HighlightingMethod

data HighlightingLevel
instance Show HighlightingLevel
instance Read HighlightingLevel


data TCEnv
data TCState
newtype TCMT m a = TCM { unTCM :: IORef TCState -> TCEnv -> m a }

instance MonadIO m => Applicative (TCMT m)
instance MonadIO m => Functor (TCMT m)
instance MonadIO m => Monad (TCMT m)
instance MonadIO m => MonadIO (TCMT m)

type TCM = TCMT IO

type ModuleToSource = Map TopLevelModuleName AbsolutePath

type BackendName = String

data Comparison
newtype ProblemId = ProblemId Nat
data Polarity
newtype CheckpointId = CheckpointId Int

class Monad m => MonadTCEnv m where
  askTC   :: m TCEnv
  localTC :: (TCEnv -> TCEnv) -> m a -> m a

  default askTC :: (MonadTrans t, MonadTCEnv n, t n ~ m) => m TCEnv
  askTC = lift askTC

  default localTC
    :: (MonadTransControl t, MonadTCEnv n, t n ~ m)
    =>  (TCEnv -> TCEnv) -> m a -> m a
  localTC = liftThrough . localTC

data TwinT'' b a
type TwinT' = TwinT'' Bool
type TwinT = TwinT' Type
data ContextHet' a
type ContextHet = ContextHet' (Dom (Name, TwinT))

data HetSide = LHS | RHS | Compat | Whole | Both
newtype Het (side :: HetSide) t = Het { unHet :: t }
instance Functor (Het side)
