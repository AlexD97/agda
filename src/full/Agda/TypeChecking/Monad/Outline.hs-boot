module Agda.TypeChecking.Monad.Outline where

import Agda.Utils.FileName
import Control.DeepSeq

data OutlineOutputCallback
data OutlinePending

jsonOutlineOutputCallback :: AbsolutePath -> OutlineOutputCallback

instance NFData OutlineOutputCallback
instance NFData OutlinePending
