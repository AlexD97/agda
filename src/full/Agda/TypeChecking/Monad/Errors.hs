module Agda.TypeChecking.Monad.Errors where

import Control.Monad.Except

import Agda.TypeChecking.Monad.MetaVars
import Agda.TypeChecking.MetaVars
import Agda.TypeChecking.Monad.Context
import Agda.TypeChecking.Monad.Base
import Agda.Syntax.Internal


import Agda.Utils.Permutation
import Agda.Utils.Size

import Agda.TypeChecking.Monad.Debug
import Agda.TypeChecking.Pretty
import Data.List.NonEmpty (NonEmpty(..))
import Agda.TypeChecking.Warnings
import Agda.Utils.Monad
import Agda.TypeChecking.Monad.Closure
import GHC.Stack (CallStack)

guardingTypeErrors :: Type -> TCM Term -> TCM Term
guardingTypeErrors want m = m `catchError` \case
  TypeError stack state err -> recoverTypeError stack state err want
  e -> throwError e

recoverTypeError :: CallStack -> TCState -> Closure TypeError -> Type -> TCM Term
recoverTypeError stack state err want = do
  i   <- createMetaInfo' DontRunMetaOccursCheck
  tel <- getContextTelescope
  m   <- newMeta' RecoveredTypeError Frozen i normalMetaPriority (idP (size tel)) $
    HasType () CmpLeq want
  reportSDoc "tc.error.recover" 10 $
    "created meta variable" <+> prettyTCM m <+> "to recover from" <+> prettyTCM err
  ifM (isErrorFromRecovery err)
    (reportSDoc "tc.error.recover" 10 $ "... but it will not become a warning")
    (do err_doc <- prettyTCM (TypeError stack state err)
        addWarning . TCWarning
          stack
          (envRange $ clEnv err)
          (RecoveredTypeErrorW (TypeError stack state err))
          err_doc =<< useR stAreWeCaching)

  MetaV m . fmap Apply <$> getContextArgs

-- | Did the given type error happen because we recovered from another
-- type error? In that case, we shouldn't add it as a warning, to
-- prevent cascading.
isErrorFromRecovery :: Closure TypeError -> TCM Bool
isErrorFromRecovery cl = enterClosure cl $ \err -> orM [errorCall] where
  isErrorMeta mid = lookupMetaInstantiation mid >>= \case
    RecoveredTypeError -> pure True
    _ -> pure False

  errorCall = asksTC (fmap clValue . envCall) >>= \case
    Just (CheckMetaSolution _ mid _ _) -> isErrorMeta mid
    _ -> pure False
