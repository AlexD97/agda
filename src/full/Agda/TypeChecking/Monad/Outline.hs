module Agda.TypeChecking.Monad.Outline where

import Agda.TypeChecking.Monad.Signature
import Agda.TypeChecking.Monad.Imports
import Agda.TypeChecking.Monad.Closure
import Agda.TypeChecking.Monad.Context
import Agda.TypeChecking.Monad.State
import Agda.TypeChecking.Monad.Debug
import Agda.TypeChecking.Monad.Base
import Agda.TypeChecking.Reduce

import {-# SOURCE #-} Agda.TypeChecking.Pretty

import qualified Agda.Interaction.Highlighting.Range as Range
import qualified Agda.Utils.RangeMap as RangeMap
import qualified Agda.Utils.Pretty as P
import Agda.Utils.Impossible
import Agda.Utils.FileName
import Agda.Utils.Lens

import Agda.Syntax.TopLevelModuleName
import Agda.Syntax.Internal.Names
import Agda.Syntax.Scope.Base
import Agda.Syntax.Position
import Agda.Syntax.Internal
import Agda.Syntax.Common

import Control.Monad.IO.Class
import Control.DeepSeq
import Control.Monad

import qualified Data.ByteString.Lazy as Bsl
import qualified Data.HashMap.Strict as HashMap
import qualified Data.Aeson.KeyMap as KeyMap
import qualified Data.Vector as Vector
import qualified Data.Text as Text
import Data.Foldable
import Data.List.NonEmpty (NonEmpty((:|)))
import Data.Monoid (Last(..))
import Data.Maybe
import Data.Aeson
import Data.Text (Text)

import GHC.Generics

import System.IO


data OutlinePending
  = OutlinePending
    { outlineRange   :: {-# UNPACK #-} !Range.Range
    , outlineContext :: Context
    , outlineScope   :: {-# UNPACK #-} !ScopeInfo
    , outlineType    :: {-# UNPACK #-} !Type
    }

data OutlineGlobal
  = OutlineGlobal
    { globalNameId :: {-# UNPACK #-} !NameId
    , globalStart  :: {-# UNPACK #-} !Int
    , globalEnd    :: {-# UNPACK #-} !Int
    , globalMod    :: {-# UNPACK #-} !TopLevelModuleName
    , globalType   :: {-# UNPACK #-} !Text
    }
  deriving (Eq, Show, Generic)

data OutlineEntry
  = OutlineEntry
    { entryStart    :: {-# UNPACK #-} !Int
    , entryEnd      :: {-# UNPACK #-} !Int
    , entryMod      :: {-# UNPACK #-} !TopLevelModuleName
    , entryContext  :: HashMap.HashMap Text Text
    , entryRelevant :: HashMap.HashMap Text OutlineGlobal
    , entryType     :: {-# UNPACK #-} !Text
    }
  deriving (Eq, Show, Generic)

instance NFData OutlineEntry
instance NFData OutlineGlobal

instance ToJSON OutlineGlobal
instance ToJSON OutlineEntry

data OutlineOutputCallback
  = OutlineOutputCallback
    { runCallback :: OutlineEntry -> TCM ()
    }

instance NFData OutlineOutputCallback where
  rnf (OutlineOutputCallback !e) = ()

instance NFData OutlinePending where
  rnf (OutlinePending r c s t) = rnf (r, c, s, t)

defaultOutlineOutputCallback :: OutlineOutputCallback
defaultOutlineOutputCallback = OutlineOutputCallback $ const $ pure ()

recordTypeInContext
  :: ( HasOptions m
     , MonadTCState m
     , MonadTCEnv m
     , ReadTCState m
     , MonadTCM m
     )
  => Range -> Type
  -> m ()
recordTypeInContext range ty = do
  ctx <- reverse <$> getContext
  scope <- getScope
  let
    pending = OutlinePending
      { outlineRange   = Range.rangeToRange range
      , outlineContext = ctx
      , outlineScope   = scope
      , outlineType    = ty
      }
  clo <- buildClosure pending
  modifyTCLens' stPendingOutline (clo:)

flushPendingOutline :: TCM ()
flushPendingOutline = do
  t <- getsTC (view stPendingOutline)
  callback <- getsTC (runCallback . stOutlineOutputCallback . stPersistentState)
  traverse_ (flip enterClosure (callback <=< finishOutline)) t
  modifyTCLens' stPendingOutline $ const []

finishOutline :: OutlinePending -> TCM OutlineEntry
finishOutline OutlinePending{..} = do
  withScope_ outlineScope $ do
    outlineContext <- runReduceM $ traverse (traverse (traverse instantiateFull')) outlineContext
    outlineType <- runReduceM $ instantiateFull' outlineType
    relevant <- flip namesIn' (fmap (fmap snd) outlineContext, outlineType) $ \nm -> do
      thedef <- getConstInfo nm
      ty <- showit (defType thedef)
      nm' <- showit nm
      let
        bs = Range.rangeToRange (nameBindingSite (qnameName (defName thedef)))
        NameId _ modh = nameId (qnameName (defName thedef))

      mod <- iTopLevelModuleName . miInterface . fromMaybe __IMPOSSIBLE__ <$>
        getVisitedModule (TopLevelModuleName noRange modh (mempty :| []))
      let
        gbl = OutlineGlobal
          { globalNameId = nameId (qnameName nm)
          , globalType   = ty
          , globalStart  = Range.from bs
          , globalEnd    = Range.to bs
          , globalMod    = mod
          }
      pure (HashMap.singleton nm' gbl)
    (ctx_e, ty) <- go outlineContext outlineType
    mod <- fromMaybe __IMPOSSIBLE__ <$> currentTopLevelModule
    pure $ OutlineEntry
      { entryStart    = Range.from outlineRange
      , entryEnd      = Range.to outlineRange
      , entryMod      = mod
      , entryType     = ty
      , entryContext  = ctx_e
      , entryRelevant = relevant
      }
  where
    showit :: PrettyTCM a => a -> TCM Text
    showit = fmap (Text.pack . P.renderStyle P.style{P.mode=P.OneLineMode}) . prettyTCM

    go :: Context -> Type -> TCM (HashMap.HashMap Text Text, Text)
    go [] t = do
      t <- showit t
      pure (mempty, t)
    go (x:xs) t = do
      let (nm, ty) = unDom x
      ty <- showit ty
      nm <- showit nm
      addContext x $ do
        (rest, t) <- go xs t
        pure (HashMap.insert nm ty rest, t)


jsonOutlineOutputCallback :: AbsolutePath -> OutlineOutputCallback
jsonOutlineOutputCallback fp =
  OutlineOutputCallback $ \obj -> do
    liftIO $ withFile (filePath fp) AppendMode $ \h -> do
      hPutChar h '\30' -- Record separator
      Bsl.hPutStr h $ encode obj
      hPutChar h '\n'
