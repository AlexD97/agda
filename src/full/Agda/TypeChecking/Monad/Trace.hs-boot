module Agda.TypeChecking.Monad.Trace where

import Data.Kind (Type)

import Agda.Syntax.Position (HasRange)
import Agda.TypeChecking.Monad.Base (TCM)

class MonadTrace (m :: Type -> Type)
instance MonadTrace TCM

setCurrentRange :: (MonadTrace m, HasRange x) => x -> m a -> m a
