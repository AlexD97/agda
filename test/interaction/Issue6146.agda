-- This file is not used other than to ensure that the test in
-- Issue6146.hs is run.
