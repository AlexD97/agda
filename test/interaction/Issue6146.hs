-- An attempt to ensure that the correct .agda-lib file is used (or an
-- error message is presented) in the presence of symbolic links and
-- .. in paths.

{-# LANGUAGE RecordWildCards #-}

import Control.Exception
import Control.Monad
import System.Directory

import RunAgda

main :: IO ()
main = runAgda [ "--ignore-interfaces"
               , "--no-default-libraries"
               ] $ \(AgdaCommands { .. }) -> do

  -- Discard the first prompt.
  echoUntilPrompt

  -- Load a file, and wait for Agda to complete.
  loadAndEcho "Issue6146/A/B/../Issue6146.agda"

  -- Create a symbolic link (and remove it afterwards).
  bracket_
    (createDirectoryLink "B/C" "Issue6146/D")
    (removeDirectoryLink "Issue6146/D") $ do

    -- Load a file, and wait for Agda to complete.
    loadAndEcho "Issue6146/D/Issue6146.agda"

  -- Cleanup.
  exists <- doesDirectoryExist "Issue6146/_build"
  when exists $ removeDirectoryRecursive "Issue6146/_build"
