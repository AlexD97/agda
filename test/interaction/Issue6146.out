(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Error*" "The module A.Issue6146 should be in the directory A/Issue6146, but Agda was given the path Issue6146/A/B/../Issue6146.agda." nil)
(agda2-highlight-load-and-delete-action)
(agda2-status-action "")
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking D.Issue6146 (Issue6146/D/Issue6146.agda). " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
